signals/src/main/java/com/vaadin/signals/impl/UsageTracker.java

```diff
@@ -0,0 +1,192 @@
+package com.vaadin.signals.impl;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import com.vaadin.signals.Id;
+import com.vaadin.signals.Node.Data;
+
+/**
+ * Tracks signal value read operations while a task is run.
+ */
+public class UsageTracker {
+    /**
+     * The type of usage. This allows a dependency to be targeted to only a
+     * specific field in {@link Data} rather than reacting to any change to that
+     * node.
+     */
+    public enum UsageType {
+        /**
+         * Usage of the regular node value.
+         */
+        VALUE(Data::value),
+
+        /**
+         * Usage of the map children of a node.
+         */
+        MAP(Data::mapChildren),
+
+        /**
+         * Usage of the list children of a node.
+         */
+        LIST(Data::listChildren),
+
+        /**
+         * Usage of the computed value. Only applicable for computed signals.
+         */
+        COMPUTED(ComputedSignal::extractRawValue),
+
+        /**
+         * Usage of any kind of data in a signal.
+         */
+        ALL(Data::lastUpdate);
+
+        private final Function<Data, Object> extractor;
+
+        private UsageType(Function<Data, Object> extractor) {
```

Comment:
Modifier 'private' is redundant for enum constructors:
```suggestion
        UsageType(Function<Data, Object> extractor) {
```